Major Update for Keyboard handling and application configuration.

DemiLovato69 opened the pull request:
For the past few days I have been working to implement config files, config file reading, and keyboard handling. 
Where we are at here: A default config file is stored as rc data in the DLL, when the program launches it copies that data to %appdata% as lightwm.config.

lightwm can then read this config file to get keybinding's from it. The keyboard handler will then register hotkeys with a dummy win32 window that will callback when a hotkey is pressed. 

Added a resource file that contains the instructions for the resource compiler to include the default lightwm.config can be used later to define an application icon and translations if needed. 
Added a def file to define exports of the dll
Modified the makefile heavily to account for this, including generating debug symbols... (when a project needs debug symbols you know how painful it was to debug)

Commit:
Added basic keyboard handling

Commit:
Added preliminary logic to read a configuration file, currently can get the appdata directory

Commit:
Added file checking and reading, now just to flesh out tokenizing

Commit:
Added ability to read config file into a config structure

Commit:
Fixed an off by one error with config items, and layed out ground work for a meta structure to hold config items

Commit:
Moved the config reader to the DLL which required reconfiguring the Makefile heavily, need to merge latest changes from remote master in order to merge feature

DemiLovato69 committed:
Merge branch 'master' into feature/keyboard

DemiLovato69 committed:
Cleanup after bringing feature to latest master changes.

DemiLovato69 committed:
Added ability to create a default config on first startup

If the lightwm.config file is missing from %appdata% then it creates a default one embedded as a resource in the dll.

DemiLovato69 committed:
A bit of cleanup and a start to the keyboard handler

DemiLovato69 committed:
Added basic actions to handle and work on the keyboard handler.

DemiLovato69 committed:
A little cleanup

Updated the ignore for CLion

Modified makefile to make pdb files for debugging

Modified keyboard handling to use RegisterHotkey win32 function

Now hotkeys are registered based on the config, the config reader needs work though since the name and value is mangled with control chars.

DemiLovato69 committed:
Major cleanup of the config reader

Need to still work out why registering meta+(any char) doesn't work

Added a debug header file to help with debug printing

Found this feature still needs work to implement actual actions but it works.

Working well now

Need to add debug flags in the makefile

DemiLovato69 committed:
Merge branch 'feature/keyboard'

DemiLovato69 committed:
Updated gitignore for ilk files

nir9 commented on config.c:
```diff
@@ -0,0 +1,294 @@
+#define BUFF_SIZE 65536
+
+/**
+ * Config reader global vars
```
(Regarding this comment and all the rest) By convention I prefer to minimize comments in the code, especially obvious ones. Only when the comment has real added value, I like adding them - but that is rare.

nir9 commented on config.c:
```diff
@@ -0,0 +1,294 @@
+/**
+ * Config Reader and Parser
+ * This program will read the config for lightwm in the appdata directory
```
No need for the description of what this file does in English words, this can be especially problamatic if someone changes the logic of the code and forgets to change the comment

I prefer to have in this case only the code as documentation :)

nir9 commented on wm_dll.c:
```diff
@@ -13,20 +23,150 @@ __declspec(dllexport) LRESULT CALLBACK ShellProc(int code, WPARAM wparam, LPARAM
+			 * Read the config file here
+			**/
+			//----------------------------------------------
+			if(LoadConfigFile(hModule) != ERROR_SUCCESS) 
+			{ 
+				return FALSE; 
+			}
+			
+			//----------------------------------------------
+			/** 
+			 * Create a hidden window to handle the hotkey messages
+			**/
+			//----------------------------------------------
+			if(!CreateHiddenWindow(hModule))
+			{
+				return FALSE; 
+			}
+			
+			//----------------------------------------------
+			/** 
+			 * Initialize the keyboard config and actions here
+			**/
+			//----------------------------------------------
+			if(!InitializeKeyboardConfig(hwnd, GetConfigItems())) 
+			{ 
+				return FALSE; 
+			}
```
This DLL is loaded into many processes on the system, therefore it is best to keep it as minimal as possible in the logic.

This logic can be moved to the main executable

nir9 commented on wm_dll.c:
```diff
@@ -13,20 +23,150 @@ __declspec(dllexport) LRESULT CALLBACK ShellProc(int code, WPARAM wparam, LPARAM
 
-BOOL APIENTRY DllMain(HMODULE hModule, DWORD ulReasonForCall, LPVOID lpReserved) 
+/**
+ * This is terrible... but I cannot think of any other way to do it with the RegisterHotkey function
```
Instead of opening a new window, how about installing another hook for key presses similar to how the shell hook is installed?

nir9 commented on wm_dll_resources.rc:
```diff
@@ -0,0 +1,88 @@
+#include "resource.h"
```
Since the DLL is loaded into many processes, we want to keep it as small as possible, therefore I would move this to the executable or just put the hardcoded default config in the code.

nir9 commented on Makefile:
```diff
@@ -1,19 +1,34 @@
 
-wm_dll.obj: wm_dll.c
-	$(CC) /c wm_dll.c
+all: $(DLL) $(EXEC)
```
```suggestion
all: clean $(DLL) $(EXEC)
```
For now also cleaning (gonna leave it like this for a couple of weeks), since I changed the EXE name and I want to save confusion of people using the old EXE by mistake.

nir9 commented on Makefile:
```diff
@@ -1,19 +1,34 @@
 
-all: clean $(DLL) $(EXEC)
+# Boiler plate
+CFLAGS  = /EHsc /Zi
```
By default, it should build as Release without the debugging flag, for debugging - I would recommend adding another rule to the make file that will build as Debug.

nir9 reviewed (changes requested):
This is very cool, thanks for the contribution! Added a few CR comments

nir9 commented on config.c:
```diff
@@ -0,0 +1,294 @@
+	
+	if(configFileHandle == NULL) 
+	{
+		SetLastError(ERROR_INVALID_HANDLE); 
```
I prefer working with last error specifically for Windows API calls, and not use it for other stuff to avoid confusion

nir9 commented on config.c:
```diff
@@ -0,0 +1,294 @@
+	if(configFileHandle == NULL) 
+	{
+		SetLastError(ERROR_INVALID_HANDLE); 
+		reportWin32Error(L"Config file could not be opened"); 
```
Since this is not a Win32 Error but a C library function that failed, I would add another function similar to reportWin32Error, but for other kinds of errors

nir9 commented on config.c:
```diff
@@ -0,0 +1,294 @@
+	}
+	
+	for(size_t lineCount = 0; fgets(line, sizeof(line), configFileHandle); lineCount++) { 
+		if(strlen(line) == 0)
```
I always use braces, even if there is only a single line, saves debugging hard to find bugs

nir9 commented on wm.c:
```diff
@@ -56,27 +66,43 @@ int main() {
 	tileWindows();
+	MSG msg; 
+	while (GetMessage(&msg, NULL, 0, 0) != 0) {
+		//TODO Need to modify this wait so the program can handle hotkeys
```
If you go with installing a keyboard hook similar to how the shell hook is installed, you can use WaitForMultipleObjects for example and work with an event for that as well, like how the window event works.

DemiLovato69 commented:
@nir9 Thanks for reviewing. I was having trouble with using a hook because there (to the best of my knowledge) isn't a hook flag to handle global hotkeys, I did get a hook to fire on a registered hotkey using an undocumented value with SetWindowsHookEx but I could not figure out how the lParam passed the data for the hotkey ID, seemed like it was either passing me garbage or some structure I don't know how to look at, if you have more info on this I would greatly appreciate it. I'll continue to try and find a better solution in the mean time. 

As for everything else I will make those changes. 😊

DemiLovato69 committed:
Merge branch 'master' into feature/keyboard

nir9 commented:
@DemiRom You can install a keyboard hook, checkout WH_KEYBOARD and WH_KEYBOARD_LL on https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw

DemiLovato69 committed:
Modified the Makefile to allow for debug builds

The debug header is also modified to only print when in debug build.

Moved resources to main exe and moved config handling to main exe

Created a custom callback to handle hotkey presses.

Fixed a bug where I was not passing the correct value to RegisterHotkey

RegisterHotkey requires a virtual key scan code not a char as I initially believed.

DemiLovato69 commented:
@nir9 I have fixed the issues you have outlined in your CR. Biggest issue I had was with the hotkeys, I ended up opting to create a custom callback procedure that gets called within the message loop when a message of type WM_HOTKEY is posted to the message queue, since both the low level and regular keyboard hook do not give you information on which registered hotkey was pressed. Without implementing custom hotkey registration logic, I wanted to keep it as simple as possible and leverage the Win32 API as much as possible, but may be a good thing to look at implementing in the future.

DemiLovato69 committed:
Forgot to uncomment the TileWindow function....

nir9 commented on wm.c:
```diff
@@ -42,41 +56,77 @@ int main() {
+			TranslateMessage(&msg); 
+			DispatchMessageW(&msg); 
+			continue;
+		} else if (WaitForSingleObject(windowEvent, INFINITE) == WAIT_FAILED) {
```
This will be blocking the message loop until the next time the event is signaled, and also the getmessage will be blocking the event from being checked.

A solution for this can be to create an new thread for handling the hot keys.

DemiLovato69 replied:
Since RegisterHotkey uses the message queue when a user presses a hotkey it puts a message into the message queue, hence why using a hook is impossible with the RegisterHotkey functionality of the Windows API. GetMessage automatically gets fired for this thread when there is a message available, and currently the only two messages that do get fired are a WM_HOTKEY message and the windowEvent message, its a simple check if the msg is of type hotkey it calls the custom callback hook in the dll with the id of the registered hotkey, then goes back to waiting for a message. 

Like I commented earlier if we resort to forgoing the RegisterHotkey and message queue functionality of the Windows API custom hotkey registration and handling functionality will have to be implemented leading to much more complexity.

nir9 replied:
The windowEvent is not sent as a message, it is a handle to an event, hence, getmessage will not return if the event is signaled and currently this logic will delay the hotkey handling or delay the window event handling (since also get message waits and also waitforsingleobject waits).

The solution for this problem (since both are currently blocking each other currently), is to do one of them in a separate thread.

nir9 replied:
BTW, no need to call DispatchMessage, there is no window procedure. Also, no need for TranslateMessage.

DemiLovato69 replied:
Sorry you are right the windowEvent doesn't post a message as I originally thought. I will rework how that gets handled, a little testing I did lead me down a rabbit hole and I thought I understood how that worked. 

As for the TranslateMessage/DispatchMessage calls, I forgot that I no longer am using a hidden window so technically neither are required.

DemiLovato69 replied:
A quick question, what was the reason to use an event instead of posting a message ?

nir9 replied:
Originally I used an event since it is simple to work with, but actually posting a thread message is a good idea in this case! (Could keep the logic simple rather than adding a new thread) - only thing is that it would require getting the thread id of the window manager in the DLL shellhook - but that could easily done by adding another function to the DLL that sets the thread id as a global variable.

nir9 commented on wm_dll.c:
```diff
@@ -13,20 +18,61 @@ __declspec(dllexport) LRESULT CALLBACK ShellProc(int code, WPARAM wparam, LPARAM
 	return CallNextHookEx(NULL, code, wparam, lparam);
 }
 
+__declspec(dllexport) LRESULT HotkeyProc(int code, WPARAM wparam, LPARAM lparam) {
```
Why have you decided to place this in the DLL? this can be in the main executable.

DemiLovato69 replied:
I was under the impression that all of the final window handling will be done in the DLL, it will have to be decided what the DLL is for, because just as easily could move the ShellProc hook to the lightwm.exe and get rid of the DLL entirely.

nir9 replied:
Since the DLL is added into many processes on the system I want to keep the logic in it as minimal as possible - that is why it is only signaling an event and the actual handling occurs in the executable.
The ShellProc can't move to the main executable and has to be in a DLL since the SetWindowsHookEx function requires it to come from a DLL if you want to place this hook system wide (see parameter lpfn and dwThreadId in https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw ).

nir9 replied:
This requirement about the hook function is basically the reason I made this DLL

DemiLovato69 replied:
If you notice in the table WH_SHELL is Thread or Global meaning it can be installed both in the installing thread or a global thread. I just made a testing branch where I removed the DLL entirely and it works just fine. 

https://github.com/DemiRom/lightwm/tree/testing/dll-removal

nir9 replied:
Try created or destroying a window after the change, I tried as well now installing the hook only on the current thread, and it doesn't catch any additional windows

DemiLovato69 replied:
Okay I see, sometimes I run it without the DLL it does work but most times it doesn't, I guess that all makes sense now. So its settled the DLL is only for the WH_SHELL hook until another global hook is required.

nir9 commented on wm_dll.c:
```diff
@@ -13,20 +18,61 @@ __declspec(dllexport) LRESULT CALLBACK ShellProc(int code, WPARAM wparam, LPARAM
+			break; 
+	}
+	
+	return CallNextHookEx(NULL, code, wparam, lparam);
```
Why are you calling CallNextHook? this function is not a hook, it is only a handler for the hot key message

DemiLovato69 replied:
Accidentally left that in since that function was originally a hook.

nir9 commented on wm_resources.rc:
```diff
@@ -0,0 +1,88 @@
+#include "resource.h"
```
To keep the code base and build as simple as possible, I think it would be better and simpler to put the default config as a text string in a header file and use it directly in the code, and remove the rc file

DemiLovato69 replied:
Absolutely could be done but with keeping the default config in a separate file as it would appear when the application puts it in the directory, I think it would make it easier to modify as features get implemented instead of modifying a string in a header. Not to mention if an icon is ever added to this executable the resource file will have to come back, the resource file can also be used to implement translations for other languages for error messages in the future.

nir9 replied:
That is indeed an upside, in that case, let's keep this resource file, but can you remove all the redundant comments and additional stuff and just leave the IDS_LIGHTWM_CNF_NAME? all the around is a lot of clutter in my eyes 😂

DemiLovato69 replied:
Absolutely can clean it up.

nir9 commented on config.c:
```diff
@@ -0,0 +1,302 @@
+	}
+}
+
+uint8_t LoadConfigFile(HINSTANCE resourceModuleHandle) 
```
As a convention I prefer the non-windows api functions to be camelCase to avoid confusion

DemiLovato69 committed:
Remove Tranlate and Dispatch Message

Clean up resources

Remove CallNextHook and return an LRESULT instead.

DemiLovato69 committed:
Remove Tranlate and Dispatch Message

Clean up resources

Remove CallNextHook and return an LRESULT instead.

Switch Config functions to camel case.

Moved macro to the keyboard header instead of the actions header.

Switch keyboard functions to camel case.

DemiLovato69 committed:
Merge branch 'feature/keyboard' of github.com:DemiRom/lightwm into feature/keyboard

DemiLovato69 committed:
Moved handling hotkeys to the keyboard

Rework how messages are handled

Found typo in Makefile for release builds

Fixed format string to handle values properly.

Reworking the config since there is a memory leak

Config works now works as expected.

Removed unused resource header include in the main application code.

GetMessage with a -1 as second param get all thread messages not window.

Cleanup comments

DemiLovato69 commented:
At this point and time I think its best I close this pull request until I figure out why the WH_SHELL hook does not work anymore, and I'm not sure why at this point. 

ShellProc only gets called when the exe thread creates or destroys a window but ignores any from windows explorer for example. 

If you have any ideas let me know.

DemiLovato69 committed:
Little makefile update

nir9 commented:
Yes, this is a little confusing, but the hook procedure runs in other executables on your system and not the window manager main executable, so when you use stuff like printf (which is how DEBUG_PRINT is implemented pretty much) - you won't see anything since the messages don't get to the console.

The solution for actually seeing messages printed from the hook procedure can be to use the OutputDebugStringW function and this sends strings to the debugger. Afterwards you can easily see these messages as they are printed live by using for example dbgview that comes from Sysinternals Suite (you can get this in the Windows store).

This is indeed a confusing thing and perhaps this is where a comment in the top of the file of the DLL with a warning about this fits.

DemiLovato69 committed:
Merge changes to remote master for fixing dpi issues.

DemiLovato69 committed:
Updated the debug header to allow to print to dbgview

Fixed format string for DWORD

Added ability to OutputDebugString to windows with formatting

Changed format to correctly read variable.

Allow the DLL to read shared memory and some cleanup.

Cleanup some TODO's

nir9 commented:
@DemiRom Have you closed by mistake?

DemiLovato69 commented:
@nir9 No, GitHub auto closed it when I reorganized my fork of the project. I will reopen it when the changes are ready. In the meantime I have got the DLL to work with your suggestion for using OutputDebugString for debugging, can't believe I didn't think of that haha, I even had the docs open for OutputDebugString that evening. In order for the DLL to get the thread ID I resorted to [Creating Named Shared Memory](https://learn.microsoft.com/en-us/windows/win32/memory/creating-named-shared-memory) which works a treat but adds a little complexity to the DLL.

nir9 commented:
@DemiRom 👍 Why not instead of shared memory, add a function to set a global variable in the DLL with the thread ID before setting the hook?

DemiLovato69 commented:
@nir9 I tried that and I might be missing something but when the dll gets loaded by other processes, it loses the thread id.

nir9 commented:
@DemiRom Oh right, that makes sense since each time the DLL is loaded into a new process, it has its own memory. In that case indeed shared memory is a good solution for transferring information.

DemiLovato69 committed:
Added workspace switching, but it's broken at the moment.

Removed workspace.c since its out of scope.

Removed all workspace handling

Not sure what to do about workspace switching.

Modified Makefile to support nmake instead of GNU make.

Added ability to compile resources.

Updated gitignore to ignore .res files.

Cleanup and added a string helper.

Reformat code

Found an off by one error in the config that was causing a crash

When cleaning the config objects configItemCount was one higher than array count for some reason.

Some basic cleanup ready for pull request.

DemiLovato69 committed:
PR #8 Keyboard Update

DemiLovato69 commented:
Much cleanup and debugging I think I have a reasonable solution now.

nir9 committed:
Merge branch 'master' into master

nir9 commented:
Awesome, I appreciate this, thanks!
